src/blockchain/compact_filters/peer.rs

```diff
@@ -57,25 +56,56 @@ pub(crate) const TIMEOUT_SECS: u64 = 30;
 #[derive(Debug, Default)]
 pub struct Mempool {
     txs: RwLock<HashMap<Txid, Transaction>>,
+    wtxids: RwLock<HashMap<Wtxid, Txid>>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+enum TxIdentifier {
+    Wtxid(Wtxid),
+    Txid(Txid),
 }
 
 impl Mempool {
+    /// Create a new empty mempool
+    pub fn new() -> Self {
+        Self::default()
+    }
+
     /// Add a transaction to the mempool
     ///
     /// Note that this doesn't propagate the transaction to other
     /// peers. To do that, [`broadcast`](crate::blockchain::Blockchain::broadcast) should be used.
     pub fn add_tx(&self, tx: Transaction) {
+        // FIXME: taking the lock separately can create race conditions, but i think it should be
```

Comment:
I think there could be a deadlock even without txid/wtxid collisions, so I think it's better to have an inner object with a single lock containing both of the HashMap to prevent this

Reply:
Ah yeah, that's a good point. I wasn't considering potential deadlocks, only potential "inconsistent" states where something is added to one map but not to the other

Reply:
now solved